src/override/memcpy.cc

```diff
@@ -79,18 +62,189 @@ namespace
    * aligned to be copied as aligned chunks of `Size` bytes.
    */
   template<size_t Size>
-  SNMALLOC_FAST_PATH bool is_aligned_memcpy(void* dst, const void* src)
+  SNMALLOC_FAST_PATH_INLINE bool is_aligned_memcpy(void* dst, const void* src)
   {
     return (pointer_align_down<Size>(const_cast<void*>(src)) == src) &&
       (pointer_align_down<Size>(dst) == dst);
   }
 
   /**
-   * Snmalloc checked memcpy.
+   * Copy a small size (`Size` bytes) as a sequence of power-of-two-sized loads
+   * and stores of decreasing size.  `Word` is the largest size to attempt for a
+   * single copy.
+   */
+  template<size_t Size, size_t Word>
+  SNMALLOC_FAST_PATH_INLINE void small_copy(void* dst, const void* src)
+  {
+    static_assert(
+      bits::next_pow2_const(Word) == Word, "Word size must be a power of two!");
+    if constexpr (Size != 0)
+    {
+      if constexpr (Size >= Word)
+      {
+        copy_one<Word>(dst, src);
+        small_copy<Size - Word, Word>(
+          pointer_offset(dst, Word), pointer_offset(src, Word));
+      }
+      else
+      {
+        small_copy<Size, Word / 2>(dst, src);
+      }
+    }
+    else
+    {
+      UNUSED(src);
+      UNUSED(dst);
+    }
+  }
+
+  /**
+   * Generate small copies for all sizes up to `Size`, using `WordSize` as the
+   * largest size to copy in a single operation.
+   */
+  template<size_t Size, size_t WordSize = Size>
+  SNMALLOC_FAST_PATH_INLINE void
+  small_copies(void* dst, const void* src, size_t len)
+  {
+    if (len == Size)
```

Comment:
Could we add an assert here that len <= Size?

Reply:
We'd end up with 16 copies of that assertion in the fast path for `memcpy`, which would make debug builds really slow.